RELEASE.txt

```diff
@@ -101,6 +105,8 @@ FIXES
 - MSVC: Erroneous construction of the installed msvc list (as described above) caused an
   index error in the msvc support code.  An explicit check was added to prevent indexing
   into an empty list.  Fixes #4312.
+- MSVC: The search for the msvc compiler executable (cl.exe) no longer inspects the
+  OS system path in certain situations when setting up the msvc environment.
```

Comment:
I don't see code change which would have this effect in this PR?
Or is this code having this effect?

```
        if not seen_path and k == 'PATH':
            seen_path = True
```

Reply:
Searching for the compiler executable using the method `find_program_path` was removed:
```
- # final check to issue a warning if the compiler is not present
-     if not find_program_path(env, 'cl'):
-         debug("did not find %s", _CL_EXE_NAME)
```

The `find_program_path` searches the OS path for the program when the program is not found in the scons environment which is undesirable in this case.   A compiler executable path from the OS PATH should not be used in this context.

This was a bug.  When looking at the existing code, one would have to know what `find_program_path` is actually doing internally to realize there is a problem.  The `find_program_path` source is shown below.

A similar issue arises when the detected dictionary keys are added to the env and *then* the environment path is searched.  Was the found compiler executable in the detection dictionary, already in the user's environment, or both?

To solve these particular issues, the detection dictionary and the user's environment are both searched immediately *prior to adding the detected path elements to the environment*:
```
    seen_path = False
    for k, v in d.items():
        if not seen_path and k == 'PATH':
            seen_path = True
            found_cl_path = SCons.Util.WhereIs('cl', v)                         # <-- DETECTED PATH
            found_cl_envpath = SCons.Util.WhereIs('cl', env['ENV'].get(k, []))  # <-- env['ENV'] PATH
        env.PrependENVPath(k, v, delete_existing=True)                          # <-- ADD DETECTED TO env['ENV']
        debug("env['ENV']['%s'] = %s", k, env['ENV'][k])
```

The code fragment Is searching the detection dictionary PATH element for a compiler executable. If no compiler executable is found in the detection dictionary this means that the user's request has failed to produce a valid compiler executable and a warning will be issued.

The code fragment is also searching the users `env['ENV']['PATH']` for a compiler executable and may append the "erroneous compiler" sentence to the end of the warning.

When there is no compiler present in the requested version detection dictionary and there is a compiler already present in the user's environment, regardless of the warning issued, the build may succeed possibly using a compiler executable that is not the same version as the request.

If if still doesn't make sense, let me know, and I'll try again.

`find_program_path` source:
```
def find_program_path(env, key_program, default_paths=None, add_path: bool=False) -> Optional[str]:
    """
    Find the location of a tool using various means.

    Mainly for windows where tools aren't all installed in /usr/bin, etc.

    Args:
        env: Current Construction Environment.
        key_program: Tool to locate.
        default_paths: List of additional paths this tool might be found in.
        add_path: If true, add path found if it was from *default_paths*.
    """
    # First search in the SCons path
    path = env.WhereIs(key_program)
    if path:
        return path

    # Then in the OS path
    path = SCons.Util.WhereIs(key_program)
    if path:
        if add_path:
            env.AppendENVPath('PATH', os.path.dirname(path))
        return path

    # Finally, add the defaults and check again.
    if default_paths is None:
        return path

    save_path = env['ENV']['PATH']
    for p in default_paths:
        env.AppendENVPath('PATH', p)
    path = env.WhereIs(key_program)

    # By default, do not change ['ENV']['PATH'] permananetly
    # leave that to the caller, unless add_path is true.
    env['ENV']['PATH'] = save_path
    if path and add_path:
        env.AppendENVPath('PATH', os.path.dirname(path))

    return path
```

Reply:
So using env['ENV']['PATH'] to find cl is only valid if MSVC_USE_SCRIPT is set to false? Otherwise you'll get the warning?

(As a side note, the variable named "d" isn't descriptive and keeps hurting my eyes.. not for this PR, but if you could name it something which indicates it's purpose at some point my eyes would thank you!  ;)

Reply:
> So using env['ENV']['PATH'] to find cl is only valid if MSVC_USE_SCRIPT is set to false? Otherwise you'll get the warning?

Not quite.

When MSVC_USE_SCRIPT is set to false this code is not executed as there is an early exit so there is never a check for a compiler.  However, a different warning is always issued.  It would probably be easy enough to suppress the warning though if a compiler was found in env['ENV']['PATH'].

Early exit:
```
...
    else:
        debug('MSVC_USE_SCRIPT set to False')
        warn_msg = "MSVC_USE_SCRIPT set to False, assuming environment " \
                   "set correctly."
        SCons.Warnings.warn(SCons.Warnings.VisualCMissingWarning, warn_msg)
        return None

  found_cl_path = None
  found_cl_envpath = None
...
```

> (As a side note, the variable named "d" isn't descriptive and keeps hurting my eyes.. not for this PR, but if you could name it something which indicates it's purpose at some point my eyes would thank you! ;)

I completely agree.

That was not my doing.  It has been that way since 2.0.0.final.0:
```
    use_script = env.get('MSVC_USE_SCRIPT', True)
    if SCons.Util.is_String(use_script):
        debug('vc.py:msvc_setup_env() use_script 1 %s\n' % repr(use_script))
        d = script_env(use_script)
    elif use_script:      
        d = msvc_find_valid_batch_script(env,version)
        debug('vc.py:msvc_setup_env() use_script 2 %s\n' % d)
        if not d:
            return d
    else:
        debug('MSVC_USE_SCRIPT set to False')
        warn_msg = "MSVC_USE_SCRIPT set to False, assuming environment " \
                   "set correctly."
        SCons.Warnings.warn(SCons.Warnings.VisualCMissingWarning, warn_msg)
        return None

    for k, v in d.items():
        debug('vc.py:msvc_setup_env() env:%s -> %s'%(k,v))
        env.PrependENVPath(k, v, delete_existing=True)
```

Reply:
And I would be happy to change the name :)